### PR TITLE
improve regex search functions signatures

### PR DESCRIPTION
**Pull request by bkamins**
This PR proposes the following changes:
* Line 170: `ismatch` should not accept any `SubString` but only `SubString{String}` as `PCRE.exec` assumes UTF-8 encoded string;
* there was a mix of two approaches in search functions: some of them accepted any `AbstractString` and converted it to `String` if needed; some of them threw an error for non-UTF-8 string; as all their docstrings indicated they should accept `AbstractString` I have made all functions accept any `AbstractString` and convert it to `String` if needed.

All the changes should be non-breaking.

### CODE REVIEW

File: base/regex.jl [hunk 1]
```diff
@@ -282,10 +278,10 @@ function matchall(re::Regex, str::String, overlap::Bool=false)
     matches
 end
 
-matchall(re::Regex, str::SubString, overlap::Bool=false) =
-    matchall(re, String(str), overlap)
+matchall(re::Regex, str::AbstractString, overlap::Bool=false) = matchall(re, String(str),
```
Comment (fredrikekre) [inline base/regex.jl]: Perhaps linebreak after `=` would look better.

Comment (bkamins) [reply base/regex.jl]: fixed

**Comment by bkamins** [comment]
additionally fixed a bug in `SubString` indexing in `match` and `matchall` via `ovec` for unicode characters (caught by Documenter build).

**Comment by bkamins** [comment]
I have also improved handling of bad `idx` along with #24103.

File: base/regex.jl [hunk 2]
```diff
@@ -214,16 +215,12 @@ function match(re::Regex, str::Union{SubString{String}, String}, idx::Integer, a
     n = div(length(ovec),2) - 1
     mat = SubString(str, ovec[1]+1, prevind(str, ovec[2]+1))
     cap = Union{Void,SubString{String}}[ovec[2i+1] == PCRE.UNSET ? nothing :
-                                        SubString(str, ovec[2i+1]+1,
-                                                  prevind(str, ovec[2i+2]+1)) for i=1:n]
+            SubString(str, ovec[2i+1]+1, prevind(str, ovec[2i+2]+1)) for i=1:n]
```
Comment (nalimilan) [inline base/regex.jl]: Incorrect alignment.

File: base/regex.jl [hunk 3]
```diff
@@ -269,7 +266,7 @@ function matchall(re::Regex, str::String, overlap::Bool=false)
             end
         end
 
-        push!(matches, SubString(str, ovec[1]+1, ovec[2]))
+        push!(matches, SubString(str, ovec[1]+1, prevind(str, ovec[2]+1)))
```
Comment (nalimilan) [inline base/regex.jl]: Can you add a comment to explain what this does? It's quite hard to guess.

File: base/regex.jl [hunk 4]
```diff
@@ -282,22 +279,20 @@ function matchall(re::Regex, str::String, overlap::Bool=false)
...
-matchall(re::Regex, str::SubString, overlap::Bool=false) =
+matchall(re::Regex, str::AbstractString, overlap::Bool=false) =
     matchall(re, String(str), overlap)
 
-function search(str::Union{String,SubString}, re::Regex, idx::Integer)
-    if idx > nextind(str,endof(str))
-        throw(BoundsError())
-    end
+function search(str::Union{String,SubString{String}}, re::Regex, idx::Integer)
+    isvalid(str, idx) || throw(ArgumentError("invalid index $i"))
```
Comment (nalimilan) [inline base/regex.jl]: "invalid character index" would be more explicit. Also throw a `UnicodeError` for consistency?

File: base/regex.jl [hunk 5]
```diff
@@ -282,22 +279,20 @@ function matchall(re::Regex, str::String, overlap::Bool=false)
...
     compile(re)
     PCRE.exec(re.regex, str, idx-1, opts, re.match_data) ?
         ((Int(re.ovec[1])+1):prevind(str,Int(re.ovec[2])+1)) : (0:-1)
 end
-search(s::AbstractString, r::Regex, idx::Integer) = throw(ArgumentError(
-    "regex search is only available for the String type; use String(s) to convert"
-))
-search(s::AbstractString, r::Regex) = search(s,r,start(s))
+search(str::Union{String,SubString{String}}, re::Regex) =
+    str == "" ? (0:-1) : search(str, re, start(s))
```
Comment (nalimilan) [inline base/regex.jl]: I don't really like this, as it means passing the index `1` is not equivalent to omitting the index. Why not add this check the the method which takes an index? Also, the current behavior had its logic, as noted at https://github.com/JuliaLang/julia/issues/24103.

Better move this to a separate PR and only keep uncontroversial changes in this one. It already contains many different things.

File: test/regex.jl [hunk 6]
```diff
@@ -34,9 +34,9 @@ show(buf, r"")
 # see #10994, #11447: PCRE2 allows NUL chars in the pattern
 @test ismatch(Regex("^a\0b\$"), "a\0b")
 
-# regex match / search string must be a String
-@test_throws ArgumentError match(r"test", GenericString("this is a test"))
-@test_throws ArgumentError search(GenericString("this is a test"), r"test")
+# regex match / search string must be a String, changed in #24116
```
Comment (nalimilan) [inline test/regex.jl]: Remove if that no longer applies.

File: base/regex.jl (hunk 2 above)
Comment (bkamins) [reply base/regex.jl]: This was the alignment used before my PR. What alignment should be correct. I have decreased in my local file the alignment to one tab so `S` in `SubString` is exactly below `=` in the above line.

File: base/regex.jl (hunk 3 above)
Comment (bkamins) [reply base/regex.jl]: You have to correct for the fact that ovec is 0-based and that ovec[2] is pointing at the last byte of the match. I am adding an appropriate comment to the source code

File: base/regex.jl (hunk 4 above)
Comment (bkamins) [reply base/regex.jl]: OK. I will separate this into `BoundsError` and `UnicodeError`. It will introduce a slight overhead though as `isvalid` does not distinguish them.

File: base/regex.jl (hunk 2 above)
Comment (nalimilan) [reply base/regex.jl]: I'd say the alignment from your PR was OK, at least it made it clear that these where inside the comprehension. (Anyway, if you didn't align everything inside the comprehension, I think it should use only four spaces, not eight; we never use tabs, so one character = one space, that's a simple rule.)

Comment (bkamins) [reply base/regex.jl]: ok - reverting to the original alignment from the PR (by tab I meant 4 spaces)

File: base/regex.jl (hunk 5 above)
Comment (bkamins) [reply base/regex.jl]: I will remove this change from this PR and add it to #24103.
I will discuss my rationale in #24103 (I have not read though it yet - I am working though this PR to fix #24157 first).
But in short passing `1` is exactly *not equivalent* to omitting the index. The corner case is an empty string "". You can search for something in an empty string (and it should not throw an error if you do not specify an index), but for this string index `1` is invalid so specifying it should be an error.

File: test/regex.jl (hunk 6 above)
Comment (bkamins) [reply test/regex.jl]: removed

**Review by bkamins — COMMENTED** [review]
I have accidentally started a review of my own code 👶.

**Comment by bkamins** [comment]
@nalimilan I am trying to locate the reason for build error (one could be allowing to pass signed integers to `PCRE.exec` as `offset` but I have to dig into it), but I want to discuss and additional thing I have spotted now.

`ismatch` takes `offset` as its third argument, on the other hand (which is not documented now) which is 0 based.
`match` takes `idx` as its third argument which is `1` based. Additionally it takes `add_opts` argument that is not accepted either by `ismatch` not `matchall` (and not all signatures of `match` allow it).

`matchall` has no `idx` nor `add_opts` argument.

I think it would be good to have it consistent. So the question is:
* should we change `offset` in `ismatch` to `idx` and make it 1-based (breaking);
* should `idx` argument be added to `matchall` (non breaking);
* shuold `add_opts` options be added to `ismatch`, `matchall` and all signatures of `match`? (non breaking); additionally I would make `add_opts` a keyword argument as someone might want to specify `add_opts` without specifying `idx` (breaking for some signatures of `match`).

**Comment by nalimilan** [comment]
Good catch. It would be good to make everything consistent indeed, even if that's breaking, since that's the last chance to fix these before 1.0. The fact that they are not documented makes it less of an issue.

Regarding `add_opts`, there's no hurry to add it to functions which do not support it since adding it won't break anything. I'm not sure about making it a keyword argument. Currently that can hurt performance, and it's easy to pass the default index using `start(s)` anyway.

BTW, note that some of these functions might be merged with generic `find` functions (see [the Search & Find Julep](https://github.com/JuliaLang/Juleps/blob/master/Find.md)). `ismatch` could be `findfirst(regexp, string) > 0`, `eachmatch` could be `findeach(regexp, string)`, etc. Though `match` is a bit special so maybe it will stay as is. So overall maybe better not make improvements to them until this is sorted out, except of course breaking changes which need to happen before 1.0.

**Comment by bkamins** [comment]
I have an additional question relating to type of `offset` used in `PCRE.exec` (I do not know whom it is best to ask this).

For instance in `regex.jl`, e.g. in `matchall` offset is defined as `UInt32`. The question is: is this a limitation of PCRE distribution that Julia uses or it might be `UInt64` or `UInt32` depending on the build of PCRE.

I ask because it is not unthinkable (or even I would say in a few years it might be normal in some types applications) to have a string larger than `typemax(UInt32)` in which case the for example the line `offset = UInt32(ovec[1]+1)` will throw `InexactError`, especially as we have a definition that `ovec::Vector{Csize_t}` and `Csize_t` does not have to be `UInt32` (and on 64 builds it will not be).

If `UInt32` is a maximum limit then I would document it. If it is not then the question is how to check what is `PCRE2_SIZE` used in PCRE build (is it guaranteed to be equal to `Csize_t`?).

File: base/regex.jl (hunk 4 above)
Comment (nalimilan) [reply base/regex.jl]: Ah, that's annoying. Make it throw either `BoundsError` or `UnicodeError`, and let's say we'll improve that later by replacing `isvalid` everywhere it's appropriate with a function which would throw the errors directly.

**Comment by nalimilan** [comment]
In `deps/build/pcre2-10.21/src/pcre2.h`, there's this:
```C
#define PCRE2_SIZE            size_t
```
So it looks like you find a bug and it should be `Csize_t`, a.k.a. `UInt` instead. Time for another PR? :-)

File: base/regex.jl (hunk 4 above)
Comment (bkamins) [reply base/regex.jl]: I reverted it to a corrected version of the original test to minimize the number of changes in this PR (as this change is breaking). I propose to implement more strict testing when your search Julep is decided.

**Comment by bkamins** [comment]
@nalimilan Given your search Julep I propose to leave this PR with:
* change `offset` to `idx` in `ismatch` and document it; it is breaking in a nasty way, as it will cause silent errors, but I believe we should unify it - I have checked Julia base and I did not find `ismatch` used with `offset` argument;
* make all `PCRE.exec` calls with `Csize_t` conversion of the offset (I decided to put it into this PR as it is related to function signatures, where we allow any `Integer`; I thought it is a better approach than restricting the signatures)
* make `match` and `matchall` accept any `AbstractString`
* correct indexing into `ovec` in `matchall`

What I did not do:
* improve bounds checking of `idx` as it is breaking and probably needs a separate PR (when a discussion of #24103 and seach Julep are decided)
* Correct use of `Csize_t` everywhere (not only in `PCRE.exec`) calls: in particular the definitions of `RegexMatch` and `Regex` are inconsistent (`Csize_t` vs `Int`), but this is a separate issue and maybe it is not critical (when limits are reached an error should be thrown, in `PCRE.exec` we have to fix it as it is a low level C call).

File: base/regex.jl [hunk 7]
```diff
@@ -154,22 +155,25 @@ r"a.a"
 julia> ismatch(rx, "aba")
 true
 
+julia> ismatch(rx, "aba", 2)
+true
```
Comment (nalimilan) [inline base/regex.jl]: `false`?

File: base/regex.jl [hunk 8]
```diff
@@ -154,22 +155,25 @@ r"a.a"
...
 
 julia> rx("aba")
 true
 ```
 """
-function ismatch(r::Regex, s::AbstractString, offset::Integer=0)
+function ismatch(r::Regex, s::AbstractString, idx::Integer=start(s))
     compile(r)
-    return PCRE.exec(r.regex, String(s), offset, r.match_options,
+    return PCRE.exec(r.regex, String(s), Csize_t(idx-1), r.match_options,
```
Comment (nalimilan) [inline base/regex.jl]: `Csize_t` is a C implementation detail, better keep it inside `PCRE.exec` as much as possible. Can't you adapt that function directly?

File: test/regex.jl [hunk 9]
```diff
@@ -34,9 +37,10 @@ show(buf, r"")
 # see #10994, #11447: PCRE2 allows NUL chars in the pattern
 @test ismatch(Regex("^a\0b\$"), "a\0b")
 
-# regex match / search string must be a String
-@test_throws ArgumentError match(r"test", GenericString("this is a test"))
```
Comment (nalimilan) [inline test/regex.jl]: If these are now supposed to work, better keep them and check that they return the correct result.

File: base/regex.jl (hunk 8 above)
Comment (nalimilan) [reply base/regex.jl]: Actually, is any change really needed? `PCRE.exec` uses `ccall`, which will take care of converting `offset` to `Csize_t`. There's no need to make the conversion manually.

Comment (bkamins) [reply base/regex.jl]: You are right. I thought, mistakenly understanding the documentation of `ccall`, that the conversion is unsafe, but it is actually safe. Removing.

File: test/regex.jl (hunk 9 above)
Comment (bkamins) [reply test/regex.jl]: fixed

File: base/regex.jl (hunk 3 above)
Comment (nalimilan) [reply base/regex.jl]: You've added the comment above, but not here.

Comment (bkamins) [reply base/regex.jl]: This issue is in three places in this file. Each follows exactly the same rule. I thought to add it to the first occurrence from top of the file.
Do you think it is worth to add it everywhere?

Comment (nalimilan) [reply base/regex.jl]: Mmm, tough call, but given how subtle this is, I'd rather be too verbose than not enough.

Comment (bkamins) [reply base/regex.jl]: ok - adding additional comment as it will not hurt

**Review by nalimilan — DISMISSED** [review]

**Comment by bkamins** [comment]
fixed

**Review by nalimilan — APPROVED** [review]

**Comment by nalimilan** [comment]
Can we merge this? It fixes a bug in `matchall` which makes CI fail for CategoricalArrays.

**Comment by bkamins** [comment]
Any objections to merge this?

**Comment by StefanKarpinski** [comment]
Even though the offset argument was undocumented, I don't think we should just change its meaning without a deprecation like this – it will silently break people's code. I'm also not even 100% on exposing the old `offset` or the new `idx` arguments at all since once can accomplish the same effect with a `SubString` without complicating the API (and very soon this will be zero-cost).

So the alternative would be to make the methods with the offset/index arguments internal (e.g. `_match`) and deprecate the undocumented offset methods to use `SubString` – in case anyone is currently using them. Does that seem reasonable?

**Comment by nalimilan** [comment]
Makes sense, anyway `ismatch` should be combined with `contains` (https://github.com/JuliaLang/julia/issues/19250). At this stage better simplify APIs, we can always add arguments later, but we won't be able to remove them.

**Comment by nalimilan** [comment]
Do you want to update this now that `ismatch` has been merged with `contain`? We really have to fix #24157 before 0.7.

**Comment by bkamins** [comment]
Rebased and readjusted (we had some major changes in the meantime in strings so I hope I have not mixed up anything 😄). I when CI passes a review is required.

Going back to the old changes I have one thing I am not 100% sure of (it was earlier used, but I want to make sure that this is correct - I have checked that it works OK, but I do not understand why):

```
function contains(s::SubString{String}, r::Regex, idx::Integer=firstindex(s))
    compile(r)
    return PCRE.exec(r.regex, s, idx-1, r.match_options,
                     r.match_data)
end
```

passes `SubString{String}` to `pcre2_match_8` from `PCRE_LIB` as `Ptr{UInt8}`. How does Julia resolve the correct pointer to the start of `SubString{String}` in this call?

File: test/regex.jl [hunk 10]
```diff
@@ -21,6 +21,9 @@ for f in [matchall, collect_eachmatch]
     @test f(r"GCG","GCGCG",overlap=true) == ["GCG","GCG"]
 end
 
+# Issue #24157
+ +@test matchall(r"fé", "café") == ["fé"]
```
Comment (nalimilan) [inline test/regex.jl]: Typo.

**Comment by nalimilan** [comment]
Thanks. Though reading @StefanKarpinski's comment above, looks like we should not document the `idx` argument, but instead deprecate it in favor of using `SubString`.

Regarding the pointer question, I guess `PCRE.exec` takes care of the conversion, or maybe `ccall` does that thanks to a `cconvert` method?

File: test/regex.jl (hunk 10 above)
Comment (bkamins) [reply test/regex.jl]: fixed

**Comment by bkamins** [comment]
`idx` - I have forgotten about that. I will deprecate `offset`. Actually it even more sense as `offset/idx` could be an invalid index and lead to weird results. Should I put the deprecation in *deprecated.jl*?
`offset` is not used in Base anywhere so this should not be a problem. I will change it when I know if CI passes

As for the `SubString{String}` issue - the reason why I have asked as `cconvert` is no-op for `SubString{String}`. Probably `ccall` does the job, but I do not understand how.

**Comment by bkamins** [comment]
CI passes so this should be good for a review.

File: base/deprecated.jl [hunk 11]
```diff
@@ -1240,6 +1240,14 @@ end
 
 @deprecate ismatch(r::Regex, s::AbstractString) contains(s, r)
 
+function ismatch(r::Regex, s::AbstractString, offset::Integer)
```
Comment (nalimilan) [inline base/deprecated.jl]: Wouldn't `@deprecate ismatch(r::Regex, s::AbstractString, offset::Integer) contains(SubString(s, offset+1), r)` be enough?

Comment (bkamins) [reply base/deprecated.jl]: it would :). Fixing

**Comment by bkamins** [comment]
This is so outdated that I close it.
